src/storage/v2/inmemory/replication/replication_client.cpp

```diff
@@ -0,0 +1,566 @@
+    : ReplicationClient{std::move(name), std::move(endpoint), mode, config}, storage_{storage} {}
+
+void InMemoryReplicationClient::TryInitializeClientAsync() {
+  thread_pool_.AddTask([this] {
+    rpc_client_.Abort();
+    this->TryInitializeClientSync();
+  });
+}
+void InMemoryReplicationClient::FrequentCheck() {
+  const auto is_success = std::invoke([this]() {
+    try {
+      auto stream{rpc_client_.Stream<replication::FrequentHeartbeatRpc>()};
+      const auto response = stream.AwaitResponse();
+      return response.success;
+    } catch (const rpc::RpcFailedException &) {
+      return false;
+    }
+  });
+  // States: READY, REPLICATING, RECOVERY, INVALID
+  // If success && ready, replicating, recovery -> stay the same because something good is going on.
+  // If success && INVALID -> [it's possible that replica came back to life] -> TryInitializeClient.
+  // If fail -> [replica is not reachable at all] -> INVALID state.
+  // NOTE: TryInitializeClient might return nothing if there is a branching point.
+  // NOTE: The early return pattern simplified the code, but the behavior should be as explained.
+  if (!is_success) {
+    replica_state_.store(replication::ReplicaState::INVALID);
+    return;
+  }
+  if (replica_state_.load() == replication::ReplicaState::INVALID) {
+    TryInitializeClientAsync();
+  }
+}  /// @throws rpc::RpcFailedException
+
+void InMemoryReplicationClient::Start() {
+  auto const &endpoint = rpc_client_.Endpoint();
+  spdlog::trace("Replication client started at: {}:{}", endpoint.address, endpoint.port);
+
+  TryInitializeClientSync();
+
+  // Help the user to get the most accurate replica state possible.
+  if (replica_check_frequency_ > std::chrono::seconds(0)) {
+    replica_checker_.Run("Replica Checker", replica_check_frequency_, [this] { FrequentCheck(); });
+  }
+}
+
+void InMemoryReplicationClient::InitializeClient() {
+  uint64_t current_commit_timestamp{kTimestampInitialId};
+
+  const auto &main_epoch = storage_->replication_state_.GetEpoch();
+
+  auto stream{rpc_client_.Stream<replication::HeartbeatRpc>(storage_->replication_state_.last_commit_timestamp_,
+                                                            main_epoch.id)};
+
+  const auto replica = stream.AwaitResponse();
+  std::optional<uint64_t> branching_point;
+  if (replica.epoch_id != main_epoch.id && replica.current_commit_timestamp != kTimestampInitialId) {
+    auto const &history = storage_->replication_state_.history;
+    const auto epoch_info_iter = std::find_if(history.crbegin(), history.crend(), [&](const auto &main_epoch_info) {
+      return main_epoch_info.first == replica.epoch_id;
+    });
+    if (epoch_info_iter == history.crend()) {
+      branching_point = 0;
+    } else if (epoch_info_iter->second != replica.current_commit_timestamp) {
+      branching_point = epoch_info_iter->second;
+    }
+  }
+  if (branching_point) {
```

Comment:
In general, I'm ok with this kind of idiomatic implicit conversion from `std::optional` to `bool`. I get in general implicit conversions are bad...but this IMO is an exception to the rule. Do you have a compelling rational why I should change my view on this?

Reply:
I think in general is ok but I don't like std::optional in combination with int since it may be confusing whether you are doing check on optional or whether it is != 0.